teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -46,17 +46,48 @@ keep track of.
 
 Today, most users view Tekton Run history via the Tekton Pipeline API (e.g.
 kubectl, tkn, dashboard, etc). However since all CRD objects are
-[stored in etcd](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/),
-this creates a problem for the Pipeline controller - as the build history grows
-over time, so does the corpus of data the controller needs to handle in its data
-plane. Over time this grows unbounded, significantly degrading the performance
-of the cluster as more Runs are completed.
+[stored in the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/),
+the data plane that is used for managing Runs by the controller is also the
+source of data users use for historical data.
+
+Over time this grows unbounded as more Runs are completed. For Tekton, this is
+particularly aggravated due to the automated nature of Runs being created.
+
+This has increased Pipeline controller latency due to `List` calls made:
+
+- [TaskRuns](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/pipelinerun/pipelinerun.go#L991)
+- [PipelineRuns](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/pipelinerun/pipelinerun.go#L998)
+- [Pods](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/taskrun/taskrun.go#L392)
+
+These latencies can also be noticed by end user tools like `kubectl get`, and
+can cause disruption in
+[extreme cases](https://github.com/kubernetes/kubectl/issues/866).
+
+In an informal stress test with a GKE cluster with ~24k TaskRuns, we observed
+`kubectl create -f taskrun.yaml` taking ~1s to complete, and
+`kubectl get taskruns` taking ~9s. A
```

Comment:
did you also observe the results if the taskruns were deleted after they completed?

Reply:
I haven't deleted them yet so I can keep the TaskRuns around for more testing if needed 🤣  If you have ideas for how we can synthetically create tons of completed TaskRuns w/o pods, let me know! (I wasn't sure if you can just create completed TaskRuns or something similar)

Anecdotally, in a cluster of mine running automated probes that experienced problems, `kubectl delete tr --all` returned the cluster to normal (this was why we originally suspected this was load related).

Reply:
> If you have ideas for how we can synthetically create tons of completed TaskRuns w/o pods, let me know!

I don't totally understand what you mean - i think you're saying that creating the test conditions takes a long time so you don't want to have to recreate it? you could update the tekton controller to not create the pods; i.e. just test the behavior of the CRDs alone

Reply:
I haven't tried it yet, but I'm curious if the controller would let me create Runs in a completed state as is. 🤔 

Though worst case, +1 to your suggesting for running a modified controller.